crates/viewer/re_ui/src/syntax_highlighting.rs

```diff
@@ -6,36 +6,74 @@ use re_log_types::{
     },
 };
 
-use crate::DesignTokens;
-use egui::{Color32, Style, TextFormat, text::LayoutJob};
+use crate::HasDesignTokens as _;
+use egui::{Color32, Style, TextFormat, TextStyle, text::LayoutJob};
 
 // ----------------------------------------------------------------------------
 pub trait SyntaxHighlighting {
     fn syntax_highlighted(&self, style: &Style) -> LayoutJob {
-        let mut builder = SyntaxHighlightedBuilder::new(style);
+        let mut builder = SyntaxHighlightedBuilder::new();
         self.syntax_highlight_into(&mut builder);
-        builder.job
+        builder.into_job(style)
     }
 
-    fn syntax_highlight_into(&self, builder: &mut SyntaxHighlightedBuilder<'_>);
+    fn syntax_highlight_into(&self, builder: &mut SyntaxHighlightedBuilder);
 }
 
 // ----------------------------------------------------------------------------
 
+enum SyntaxHighlightedStyle {
+    StringValue,
+    Identifier,
+    Index,
+    Primitive,
+    Syntax,
+    Body,
+    BodyItalics,
+    Custom(Box<TextFormat>),
+    CustomClosure(Box<dyn Fn(&Style) -> TextFormat>),
+}
+
+struct SyntaxHighlightedPart {
+    byte_range: std::ops::Range<usize>,
+    style: SyntaxHighlightedStyle,
+}
+
 /// Easily build syntax-highlighted text.
-pub struct SyntaxHighlightedBuilder<'a> {
-    pub style: &'a Style,
-    pub tokens: &'a DesignTokens,
-    pub job: LayoutJob,
+pub struct SyntaxHighlightedBuilder {
+    text: String,
+    parts: Vec<SyntaxHighlightedPart>,
```

Comment:
We could consider using a `SmallVec` here since most things with syntax-highlighting are rather short